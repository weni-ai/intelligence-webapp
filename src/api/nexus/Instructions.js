import nexusRequest from '../nexusaiRequest';
import { InstructionAdapter } from '../adapters/instructions/instruction';
import { useInstructionsStore } from '@/store/Instructions';

const request = nexusRequest;

export const Instructions = {
  async addInstruction({ projectUuid, instruction }) {
    const instructionsStore = useInstructionsStore();
    const body = {
      instructions: [...instructionsStore.instructions.data, instruction].map(
        InstructionAdapter.toApi,
      ),
    };
    const response = await request.$http.put(
      `api/${projectUuid}/customization/`,
      body,
      {
        hideGenericErrorAlert: true,
      },
    );

    const biggestId = response.data.instructions.reduce((acc, instruction) => {
      if (instruction.id > acc) acc = instruction.id;
      return acc;
    }, 0);

    return InstructionAdapter.fromApi(
      response.data.instructions.find(
        (instruction) => instruction.id === biggestId,
      ),
    );
  },

  async list({ projectUuid }) {
    const response = await request.$http.get(
      `api/${projectUuid}/customization/`,
    );

    return response.data.instructions.map(InstructionAdapter.fromApi);
  },

  async edit({ projectUuid, id, text }) {
    const instructionsStore = useInstructionsStore();

    const body = {
      instructions: [
        ...instructionsStore.instructions.data.filter(
          (instruction) => instruction.id !== id,
        ),
        { id, text },
      ].map(InstructionAdapter.toApi),
    };
    await request.$http.put(`api/${projectUuid}/customization/`, body);
  },

  async delete({ projectUuid, id }) {
    await request.$http.delete(`api/${projectUuid}/customization/?id=${id}`);
  },
};
