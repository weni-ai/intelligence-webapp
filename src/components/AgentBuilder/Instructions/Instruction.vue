<template>
  <section class="instruction">
    <p
      class="instruction__text"
      data-testid="instruction-text"
    >
      {{ text }}
    </p>

    <p
      v-if="tag"
      class="instruction__tag"
      data-testid="instruction-tag"
    >
      {{ tag }}
    </p>
  </section>
</template>

<script setup>
const props = defineProps({
  text: {
    type: String,
    required: true,
  },
  tag: {
    type: String,
    default: '',
    required: false,
  },
});
</script>

<style lang="scss" scoped>
.instruction {
  display: flex;
  justify-content: space-between;
  align-items: center;
  gap: $unnnic-spacing-sm;

  padding: $unnnic-spacing-sm;

  &:not(:last-child) {
    border-bottom: $unnnic-border-width-thinner solid $unnnic-color-neutral-soft;
  }

  &__text {
    margin: 0;
    color: $unnnic-color-neutral-cloudy;

    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-gt;
    font-weight: $unnnic-font-weight-regular;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
  }

  &__tag {
    border-radius: $unnnic-border-radius-pill;

    padding: $unnnic-spacing-nano $unnnic-spacing-ant;

    background-color: $unnnic-color-neutral-light;
    color: $unnnic-color-neutral-darkest;
    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-md;
    font-weight: $unnnic-font-weight-regular;
    line-height: $unnnic-font-size-body-md + $unnnic-line-height-md;
  }
}
</style>
