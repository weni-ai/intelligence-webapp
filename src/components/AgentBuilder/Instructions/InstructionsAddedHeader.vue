<template>
  <header class="instructions-added-header">
    <hgroup
      class="instructions-added-header__infos"
      data-testid="instructions-header-infos"
    >
      <h2
        class="infos__title"
        data-testid="instructions-title"
      >
        {{ $t('agent_builder.instructions.instructions_added.title') }}
      </h2>
      <h3
        class="infos__subtitle"
        data-testid="instructions-subtitle"
      >
        {{
          $t('agent_builder.instructions.instructions_added.subtitle', {
            count: instructionsCount,
          })
        }}
      </h3>
    </hgroup>

    <UnnnicSelectSmart
      v-model="filters"
      :options="filtersOptions"
      orderedByIndex
      multiple
      multipleWithoutSelectsMessage
      autocomplete
      data-testid="instructions-filters"
    />
  </header>
</template>

<script setup>
import { computed } from 'vue';
import i18n from '@/utils/plugins/i18n';

defineProps({
  instructionsCount: {
    type: Number,
    required: true,
  },
});

const filters = defineModel('filters', {
  type: Array,
  required: true,
});

const getFilterTranslation = (filter) =>
  i18n.global.t(
    `agent_builder.instructions.instructions_added.filters.${filter}`,
  );
const filtersOptions = computed(() => [
  {
    label: getFilterTranslation('placeholder'),
    value: '',
  },
  {
    label: getFilterTranslation('default_instructions'),
    value: 'default_instructions',
  },
  {
    label: getFilterTranslation('personalized_instructions'),
    value: 'personalized_instructions',
  },
]);
</script>

<style lang="scss" scoped>
.instructions-added-header {
  display: grid;
  grid-template-columns: 9fr 3fr;
  align-items: center;
  gap: $unnnic-spacing-sm;

  &__infos {
    display: flex;
    flex-direction: column;
    gap: $unnnic-spacing-nano;

    .infos__title {
      margin: 0;

      color: $unnnic-color-neutral-darkest;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-lg;
      font-weight: $unnnic-font-weight-bold;
      line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
    }

    .infos__subtitle {
      margin: 0;

      color: $unnnic-color-neutral-cloudy;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-md;
      font-weight: $unnnic-font-weight-regular;
      line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
    }
  }
}
</style>
