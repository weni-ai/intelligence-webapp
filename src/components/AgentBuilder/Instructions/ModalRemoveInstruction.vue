<template>
  <UnnnicModalDialog
    :modelValue="modelValue"
    data-testid="modal"
    class="modal-remove-instruction"
    showCloseIcon
    icon="warning"
    iconScheme="aux-red-500"
    :title="$t('agent_builder.instructions.remove_instruction.title')"
    size="sm"
    :secondaryButtonProps="{
      text: $t('agent_builder.instructions.remove_instruction.cancel'),
    }"
    :primaryButtonProps="{
      text: $t('agent_builder.instructions.remove_instruction.remove'),
      loading: instruction.status === 'loading',
      type: 'warning',
    }"
    @secondary-button-click="close"
    @primary-button-click="removeInstruction"
    @update:model-value="close"
  >
    <p
      class="modal-remove-instruction__description"
      data-testid="description"
    >
      {{
        $t('agent_builder.instructions.remove_instruction.modal_description')
      }}
    </p>
  </UnnnicModalDialog>
</template>

<script setup>
import { useInstructionsStore } from '@/store/Instructions';
import { ref } from 'vue';

const instructionsStore = useInstructionsStore();

const emit = defineEmits(['update:modelValue']);

const props = defineProps({
  instruction: {
    type: Object,
    required: true,
  },
});

const modelValue = ref(false);

function close() {
  emit('update:modelValue', false);
}

async function removeInstruction() {
  const { status } = await instructionsStore.removeInstruction(
    props.instruction.id,
  );

  if (status === 'complete') close();
}
</script>

<style lang="scss" scoped>
.modal-remove-instruction {
  :deep(.unnnic-modal-dialog__container__content) {
    padding-bottom: $unnnic-spacing-xs;
  }

  &__description {
    color: $unnnic-color-neutral-cloudy;

    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-gt;
    font-weight: $unnnic-font-weight-regular;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
  }
}
</style>
