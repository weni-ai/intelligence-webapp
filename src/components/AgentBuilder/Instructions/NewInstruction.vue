<template>
  <section
    class="new-instruction"
    data-testid="new-instruction"
  >
    <h2
      class="new-instruction__title"
      data-testid="new-instruction-title"
    >
      {{ $t('agent_builder.instructions.new_instruction.title') }}
    </h2>
    <UnnnicTextArea
      v-model="instructionsStore.newInstruction.text"
      data-testid="new-instruction-textarea"
      :placeholder="
        $t('agent_builder.instructions.new_instruction.textarea.placeholder')
      "
      :message="
        $t('agent_builder.instructions.new_instruction.textarea.description')
      "
    />
    <UnnnicButton
      class="new-instruction__add-instruction-button"
      data-testid="add-instruction-button"
      :disabled="!newInstruction.text.trim()"
      :text="$t('agent_builder.instructions.new_instruction.add_instruction')"
      :loading="newInstruction.status === 'loading'"
      @click="instructionsStore.addInstruction"
    />
  </section>
</template>

<script setup>
import { computed } from 'vue';
import { useInstructionsStore } from '@/store/Instructions';

const instructionsStore = useInstructionsStore();
const newInstruction = computed(() => instructionsStore.newInstruction);
</script>

<style lang="scss" scoped>
.new-instruction {
  border-radius: $unnnic-border-radius-md;
  border: $unnnic-border-width-thinner solid $unnnic-color-neutral-soft;
  padding: $unnnic-spacing-sm;

  display: grid;
  grid-template-columns: repeat(12, 1fr);
  row-gap: $unnnic-spacing-sm;

  & > * {
    grid-column: 1 / -1;
  }

  &__title {
    margin: 0;

    color: $unnnic-color-neutral-darkest;
    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-lg;
    font-weight: $unnnic-font-weight-bold;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
  }

  &__add-instruction-button {
    grid-column: 1 / 5;
  }
}
</style>
