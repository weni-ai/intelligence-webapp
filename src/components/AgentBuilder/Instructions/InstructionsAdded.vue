<template>
  <section class="instructions-added">
    <h2
      class="instructions-added__title"
      data-testid="instructions-title"
    >
      {{ $t('agent_builder.instructions.instructions_added.title') }}
    </h2>

    <section
      :class="[
        'instructions-added__instructions',
        {
          'instructions-added__instructions--loading':
            instructionsStore.instructions.status === 'loading',
        },
      ]"
      data-testid="instructions"
    >
      <template v-if="instructionsStore.instructions.status === 'loading'">
        <UnnnicSkeletonLoading
          v-for="index in 10"
          :key="index"
          tag="div"
          width="100%"
          height="60px"
          data-testid="instruction-loading"
        />
      </template>
      <template v-else>
        <Instruction
          v-for="(instruction, index) in instructionsDefault"
          :key="index"
          data-testid="instruction-default"
          :text="instruction"
          :tag="
            $t(
              'agent_builder.instructions.instructions_added.default_instruction',
            )
          "
        />
        <Instruction
          v-for="instruction in instructionsStore.instructions.data"
          :key="instruction.id"
          data-testid="instruction-added"
          :text="instruction.text"
        />
      </template>
    </section>
  </section>
</template>

<script setup>
import { computed } from 'vue';
import { useInstructionsStore } from '@/store/Instructions';

import Instruction from './Instruction.vue';
import i18n from '@/utils/plugins/i18n';

const instructionsStore = useInstructionsStore();
if (instructionsStore.instructions.status === null) {
  instructionsStore.loadInstructions();
}

const instructionsDefault = computed(() =>
  i18n.global.tm(
    'agent_builder.instructions.instructions_added.default_instructions',
  ),
);
</script>

<style lang="scss" scoped>
.instructions-added {
  display: flex;
  flex-direction: column;
  gap: $unnnic-spacing-sm;

  &__title {
    margin: 0;

    color: $unnnic-color-neutral-darkest;
    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-lg;
    font-weight: $unnnic-font-weight-bold;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
  }

  &__instructions {
    display: flex;
    flex-direction: column;

    border-radius: $unnnic-border-radius-md;
    border: $unnnic-border-width-thinner solid $unnnic-color-neutral-soft;

    &--loading {
      gap: $unnnic-spacing-nano;

      border: none;
      border-radius: 0;
    }
  }
}
</style>
