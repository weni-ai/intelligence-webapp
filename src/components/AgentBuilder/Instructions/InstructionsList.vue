<template>
  <section
    :class="[
      'instructions-list',
      {
        'instructions-list--loading': isLoading,
        'instructions-list--no-instructions': instructions.length === 0,
      },
    ]"
    data-testid="instructions"
  >
    <template v-if="isLoading">
      <UnnnicSkeletonLoading
        v-for="index in 10"
        :key="index"
        tag="div"
        width="100%"
        height="60px"
        data-testid="instruction-loading"
      />
    </template>
    <section
      v-else-if="instructions.length === 0"
      class="instructions-list__no-instructions"
    >
      <UnnnicIcon
        icon="format_list_bulleted"
        size="xl"
        scheme="neutral-cleanest"
      />
      <p class="no-instructions__text">
        {{
          $t(
            'agent_builder.instructions.instructions_list.no_custom_instructions',
          )
        }}
      </p>
    </section>
    <template v-else>
      <Instruction
        v-for="instruction in instructions"
        :key="instruction.id"
        :instruction="instruction"
        :showActions="showActions"
        data-testid="instruction-added"
      />
    </template>
  </section>
</template>

<script setup>
import Instruction from './Instruction.vue';

defineProps({
  instructions: {
    type: Array,
    required: true,
  },
  isLoading: {
    type: Boolean,
    default: false,
  },
  showActions: {
    type: Boolean,
    default: false,
  },
});
</script>

<style lang="scss" scoped>
.instructions-list {
  margin-bottom: $unnnic-spacing-sm;

  display: flex;
  flex-direction: column;

  border-radius: $unnnic-border-radius-md;
  border: $unnnic-border-width-thinner solid $unnnic-color-neutral-soft;

  &--loading,
  &--no-instructions {
    height: 100%;

    gap: $unnnic-spacing-nano;

    border: none;
    border-radius: 0;
  }

  &__no-instructions {
    height: 100%;

    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    gap: $unnnic-spacing-nano;

    .no-instructions__text {
      color: $unnnic-color-neutral-cloudy;
      font-family: $unnnic-font-family-secondary;
      font-size: $unnnic-font-size-body-gt;
      font-weight: $unnnic-font-weight-regular;
      line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
    }
  }
}
</style>
