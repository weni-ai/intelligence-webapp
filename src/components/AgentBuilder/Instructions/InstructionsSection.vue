<template>
  <section class="instructions-section">
    <h2
      class="instructions-section__title"
      data-testid="instructions-section-title"
    >
      {{ $t('agent_builder.instructions.instructions_list.title') }}
    </h2>

    <UnnnicTab
      :tabs="tabs"
      :activeTab="activeTab"
      data-testid="instructions-section-tab"
      @change="instructionsStore.activeInstructionsListTab = $event"
    >
      <template
        v-for="tab in tabs"
        :key="tab"
        #[`tab-head-${tab}`]
      >
        {{ $t(`agent_builder.instructions.instructions_list.tabs.${tab}`) }}
      </template>
      <template #[`tab-panel-${activeTab}`]>
        <p class="instructions-section__tab-description">
          {{
            $t(
              `agent_builder.instructions.instructions_list.tabs.${activeTab}_description`,
            )
          }}
        </p>
      </template>
    </UnnnicTab>
    <InstructionsList
      v-if="activeTab === 'custom'"
      data-testid="instructions-custom"
      :instructions="instructionsStore.instructions.data"
      :isLoading="instructionsStore.instructions.status === 'loading'"
      showActions
    />
    <InstructionsList
      v-if="activeTab === 'default'"
      data-testid="instructions-default"
      :instructions="instructionsDefault"
      :isLoading="false"
      :showActions="false"
    />
    <InstructionsList
      v-if="activeTab === 'safety_topics'"
      data-testid="instructions-safety-topics"
      :instructions="instructionsSafetyTopics"
      :isLoading="false"
      :showActions="false"
    />
  </section>
</template>

<script setup>
import { computed, ref } from 'vue';
import { useInstructionsStore } from '@/store/Instructions';

import InstructionsList from './InstructionsList.vue';
import i18n from '@/utils/plugins/i18n';

const instructionsStore = useInstructionsStore();

if (instructionsStore.instructions.status === null) {
  instructionsStore.loadInstructions();
}

const tabs = ref(['custom', 'default', 'safety_topics']);
const activeTab = computed(() => instructionsStore.activeInstructionsListTab);

function generateMockedInstructions(type) {
  return i18n.global
    .tm(`agent_builder.instructions.instructions_list.${type}`)
    .map((instruction, index) => ({
      id: `${type}-${index}`,
      text: instruction,
    }));
}

const instructionsDefault = computed(() =>
  generateMockedInstructions('default_instructions'),
);

const instructionsSafetyTopics = computed(() =>
  generateMockedInstructions('safety_topics'),
);
</script>

<style lang="scss" scoped>
.instructions-section {
  display: flex;
  flex-direction: column;

  &__title {
    margin: 0 0 $unnnic-spacing-sm;

    color: $unnnic-color-neutral-darkest;
    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-lg;
    font-weight: $unnnic-font-weight-bold;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
  }

  &__tab-description {
    margin-bottom: $unnnic-spacing-xs;

    color: $unnnic-color-neutral-cloudy;

    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-body-gt;
    font-weight: $unnnic-font-weight-regular;
    line-height: $unnnic-font-size-body-gt + $unnnic-line-height-md;
  }
}
</style>
