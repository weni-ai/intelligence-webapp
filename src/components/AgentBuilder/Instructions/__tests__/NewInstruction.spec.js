import { shallowMount } from '@vue/test-utils';
import { describe, it, expect, beforeEach } from 'vitest';

import NewInstruction from '../NewInstruction.vue';
import i18n from '@/utils/plugins/i18n';
import { createTestingPinia } from '@pinia/testing';
import { useInstructionsStore } from '@/store/Instructions';
import { nextTick } from 'vue';

describe('NewInstruction.vue', () => {
  let wrapper;
  const pinia = createTestingPinia({
    initialState: {
      instructions: {
        data: [],
      },
      newInstruction: {
        text: '',
        status: null,
      },
    },
  });
  let instructionsStore;

  beforeEach(() => {
    instructionsStore = useInstructionsStore(pinia);

    wrapper = shallowMount(NewInstruction, {
      global: {
        plugins: [pinia],
      },
    });
  });

  const findComponent = (component) =>
    wrapper.findComponent(SELECTORS[component]);
  const find = (selector) => wrapper.find(SELECTORS[selector]);

  const translation = (key) =>
    i18n.global.t(`agent_builder.instructions.new_instruction.${key}`);

  const SELECTORS = {
    container: '[data-testid="new-instruction"]',
    title: '[data-testid="new-instruction-title"]',
    textarea: '[data-testid="new-instruction-textarea"]',
    addButton: '[data-testid="add-instruction-button"]',
  };

  describe('Component rendering', () => {
    it('should match snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('renders a title, description, textarea and add instruction button', () => {
      expect(find('container').exists()).toBe(true);
      expect(find('title').exists()).toBe(true);
      expect(findComponent('textarea').exists()).toBe(true);
      expect(findComponent('addButton').exists()).toBe(true);
    });

    describe('Title rendering', () => {
      it('renders the correct title', () => {
        expect(find('title').text()).toBe(translation('title'));
      });
    });

    describe('Textarea rendering', () => {
      it('provides the correct props to textarea', () => {
        expect(findComponent('textarea').props('placeholder')).toBe(
          translation('textarea.placeholder'),
        );
        expect(findComponent('textarea').props('message')).toBe(
          translation('textarea.description'),
        );
      });
    });

    describe('Add button rendering', () => {
      it('disables the button when the textarea is empty', () => {
        expect(findComponent('addButton').props('disabled')).toBe(true);
      });
    });
  });
});
