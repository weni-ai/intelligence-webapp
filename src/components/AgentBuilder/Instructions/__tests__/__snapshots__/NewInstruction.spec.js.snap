// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NewInstruction.vue > Component rendering > should match snapshot 1`] = `
"<section data-v-5fa2c1a3="" class="new-instruction" data-testid="new-instruction">
  <h2 data-v-5fa2c1a3="" class="new-instruction__title" data-testid="new-instruction-title">New custom instruction</h2>
  <anonymous-stub data-v-5fa2c1a3="" modelvalue="" placeholder="You're funny, but don't make jokes" size="md" message="Keep instructions clear, direct, and in the imperative." disabled="false" type="normal" errors="" data-testid="new-instruction-textarea"></anonymous-stub>
  <unnnic-button-stub data-v-5fa2c1a3="" size="large" text="Add instruction" type="primary" float="false" iconleft="" iconright="" iconcenter="" iconsfilled="false" next="false" disabled="true" loading="false" class="new-instruction__add-instruction-button" data-testid="add-instruction-button"></unnnic-button-stub>
</section>"
`;
